### PR TITLE
HParams: Create filter dialog component

### PR DESCRIPTION
**Pull request by rileyajones**
## Motivation for features / changes
We want to be able to filter the runs table by hparam. The existing logic for this is handled by a bunch of inner components found in the runs_table_component which makes it hard to share. In this PR I am replicating that UI with some enhancements in a standalone component which will later be used by the data_table_component.

See #6488 for screenshots and a more complete description of what is being added

## Alternate designs / implementations considered (or N/A)
I'm not sure that this component should be part of the data table but that is simplest place to put it for now.

### CODE REVIEW

File: tensorboard/webapp/hparams/_types.ts [hunk 1]
```diff
@@ -13,12 +13,12 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 import {
-  DiscreteHparamValues,
-  DomainType,
   HparamSpec,
   MetricSpec,
 } from '../runs/data_source/runs_data_source_types';
 
+export {DiscreteFilter, IntervalFilter} from '../widgets/data_table/types';
```
Comment (rileyajones) [inline tensorboard/webapp/hparams/_types.ts:20]: I don't like re-exporting things but there are several files which import and then reexoport FROM this file and I would rather wait to cleaning the whole mess up until AFTER launching HParams in Timeseries Phase 1.

File: tensorboard/webapp/runs/views/runs_table/runs_table_container.ts [hunk 2]
```diff
@@ -192,8 +192,8 @@ function sortTableDataItems(
     let bValue = b[sort.name];
 
     if (sort.name === 'experimentAlias') {
-      aValue = (aValue as ExperimentAlias).aliasNumber;
-      bValue = (bValue as ExperimentAlias).aliasNumber;
+      aValue = a.experimentAlias!.aliasNumber;
```
Comment (rileyajones) [inline tensorboard/webapp/runs/views/runs_table/runs_table_container.ts:195]: This should always be set in the runs table so the non null assertion should be fine. I don't mind adding a null check to the if statement on line 194 though 🤷.

File: tensorboard/webapp/widgets/data_table/filter_dialogue.ng.html [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    *ngIf="filter.type === DomainType.DISCRETE"
+    class="discrete-filters-area"
+  >
+    <tb-filter-input
+      *ngIf="filter.possibleValues"
+      [value]="discreteValueFilter"
+      (keyup)="discreteValueKeyUp($event)"
+      placeholder="Filter Discrete Values (regex)"
+    ></tb-filter-input>
+    <div *ngIf="!getPossibleValues().length" class="no-matches">
```
Comment (rileyajones) [inline tensorboard/webapp/widgets/data_table/filter_dialogue.ng.html:25]: I have some minor concerns around running this filter twice on every render if there gets to be a large number of options but I would like to find a way to test that situation first before spending the time to optimize this.

File: tensorboard/webapp/widgets/data_table/BUILD [hunk 4]
```diff
@@ -123,11 +135,37 @@ tf_ng_module(
...
+        "filter_dialogue_module.ts",
+    ],
+    assets = [
+        "filter_dialogue.ng.html",
+        ":filter_dialogue_styles",
+    ],
+    deps = [
+        ":types",
+        "//tensorboard/webapp/angular:expect_angular_material_button",
+        "//tensorboard/webapp/angular:expect_angular_material_checkbox",
```
Comment (rileyajones) [inline tensorboard/webapp/widgets/data_table/BUILD:151]: Note that this will need to be added to the legacy component allowlist before being synced internally. cl/547923944

File: tensorboard/webapp/widgets/data_table/filter_dialogue.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      return values;
+    }
+    return values.filter((value) =>
+      value.toString().match(this.discreteValueFilter)
+    );
+  }
+
+  discreteValueKeyUp(event: KeyboardEvent) {
+    this.discreteValueFilter = (event.target! as HTMLInputElement).value;
+  }
```
Comment (rileyajones) [inline tensorboard/webapp/widgets/data_table/filter_dialogue.ts:55]: This is how I have seen the filter input used before but this is kinda janky. I'm considering
 * Doing this in the template to avoid a somewhat pointless function
 * Modifying the filter input component to either emit the input fields value or support the dreaded two way binding

File: tensorboard/webapp/widgets/data_table/types.ts [hunk 6]
```diff
@@ -74,8 +115,9 @@ export interface SortingInfo {
  * DataTable. It will have a value for each required ColumnHeader for a given
  * run.
  */
-export type TableData = Record<string, string | number | boolean | object> & {
+export type TableData = Record<string, string | number | boolean> & {
```
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/types.ts:118]: I'm happy breaking this out into a new PR, I did mention why this was being done in the pr description.

image

File: tensorboard/webapp/widgets/data_table/types.ts [hunk 7]
```diff
@@ -42,6 +45,44 @@ export enum ColumnHeaderType {
...
+export enum DomainType {
+  DISCRETE,
+  INTERVAL,
+}
+
+export type DiscreteFilterValues = string[] | number[] | boolean[];
+
+export type DiscreteFilterValue = DiscreteFilterValues[number];
+
+export interface DiscreteFilter {
```
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/types.ts:57]: The context menu is controlled by the data table and the filter options are controlled by the context menu.

I'll try to rework this to use generics.

File: tensorboard/webapp/widgets/data_table/types.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/types.ts:118]: #6494

File: tensorboard/webapp/widgets/data_table/types.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/webapp/widgets/data_table/types.ts:57]: Hopefully you don't need to do anything complex like that. If you're fighting a circular dependency then usually you can get away with putting the types in a third location. Maybe in this case hparams/filter_types.ts or similar. Or perhaps even data_table/hparam_types.ts.

File: tensorboard/webapp/widgets/data_table/BUILD (hunk 4 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/BUILD:151]: The styling looks a little off but I can probably overcome the issue.
image

File: tensorboard/webapp/widgets/data_table/filter_dialog.scss [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+.filter-dialog {
+  padding: 8px;
+  border-radius: 4px;
+  border: 1px solid;
+  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
+  border-color: mat.get-color-from-palette($tb-foreground, border);
+  background-color: mat.get-color-from-palette($tb-background, background);
+
+  @include tb-dark-theme {
+    border-color: mat.get-color-from-palette($tb-dark-foreground, border);
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog.scss:27]: Same, seems like this should be moved in custom dialog component, if possible.

File: tensorboard/webapp/widgets/data_table/filter_dialog.scss [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    );
+  }
+
+  .discrete-filters-area {
+    max-height: 100px;
+    overflow-y: auto;
+  }
+
+  .no-matches {
+    // 24px is the width of the checkbox so the text should match the
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog.scss:40]: For what it's worth, I played with it a bit and it didn't seem like the "No matching values" text ever "matched the indentation of the selectable filters" for any definition of "matched".

image

image

File: tensorboard/webapp/widgets/data_table/filter_dialog.ng.html [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      >
+    </div>
+  </div>
+
+  <div
+    *ngIf="filter.type === DomainType.INTERVAL"
+    (click)="$event.stopPropagation()"
+    class="range-input-container"
+    disableRipple
+    mat-menu-item
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog.ng.html:51]: Same, couldn't determine any reason why mat-menu-item needs to be here.

File: tensorboard/webapp/widgets/data_table/types.ts [hunk 11]
```diff
@@ -42,6 +43,44 @@ export enum ColumnHeaderType {
...
+  type: DomainType.DISCRETE;
+  includeUndefined: boolean;
+  possibleValues: DiscreteFilterValues;
+  // Subset of `possibleValues`
+  filterValues: DiscreteFilterValues;
+}
+
+export interface DiscreteFilterChange {
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/types.ts:63]: I do not see this type used anywhere in this PR. Can we add it when needed? (And, by the way, it includes a field "hparamName" so you may want to figure out how to make that more generic).

File: tensorboard/webapp/widgets/data_table/types.ts [hunk 12]
```diff
@@ -42,6 +43,44 @@ export enum ColumnHeaderType {
...
+  type: DomainType.INTERVAL;
+  includeUndefined: boolean;
+  minValue: number;
+  maxValue: number;
+  // Filter values have to be in between min and max values (inclusive).
+  filterLowerValue: number;
+  filterUpperValue: number;
+}
+
+export interface FilterAddedEvent {
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/types.ts:79]: Same, I do not see this type used anywhere in this PR. Can we add it when needed?

File: tensorboard/webapp/widgets/data_table/types.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/webapp/widgets/data_table/types.ts]: To close the loop, as discussed offline: Turns out the "type errors" were compile errors and were just the result of a bad merge.

File: tensorboard/webapp/widgets/data_table/filter_dialogue_test.ts [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    const includeUndefinedSpy = spyOn(
+      fixture.componentInstance.includeUndefinedToggled,
+      'emit'
+    );
+    const includeUndefinedCheckbox = Array.from(
+      fixture.debugElement.queryAll(By.css('mat-checkbox'))
+    ).find((elem) =>
+      elem.nativeElement.innerHTML.includes('Include Undefined')
+    )!;
+    includeUndefinedCheckbox.query(By.css('label')).nativeElement.click();
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/webapp/widgets/data_table/filter_dialogue_test.ts]: Did you give Component Harnesses a try?

File: tensorboard/webapp/runs/views/runs_table/runs_table_container.ts (hunk 2 above)
Comment (bmd3k) [reply tensorboard/webapp/runs/views/runs_table/runs_table_container.ts]: This is now irrelevant.

File: tensorboard/webapp/hparams/_types.ts (hunk 1 above)
Comment (rileyajones) [reply tensorboard/webapp/hparams/_types.ts:20]: I added it to our spreadsheet https://docs.google.com/spreadsheets/d/1RBodm1hj3-njiowUeRXQe3h1GlVb70Ce1JvZw_Cp_Yc/edit?resourcekey=0-ehUesPxhJhOONsW6w_SXsA#gid=0

File: tensorboard/webapp/widgets/data_table/filter_dialogue_test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialogue_test.ts]: I did manage to get this working with harnesses

File: tensorboard/webapp/widgets/data_table/filter_dialog.ng.html [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+      [value]="discreteValueFilter"
+      (keyup)="discreteValueKeyUp($event)"
+      placeholder="Filter Discrete Values (regex)"
+    ></tb-filter-input>
+    <div *ngIf="!getPossibleValues().length" class="no-matches">
+      No Matching Values
+    </div>
+    <div
+      *ngFor="let value of getPossibleValues(); index"
+      mat-menu-item
```
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialog.ng.html:33]: I think these were holdovers from when I copied this out from the runs table component that I just missed cleaning up. I tested and they are not needed.

File: tensorboard/webapp/widgets/data_table/types.ts (hunk 11 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/types.ts:63]: Removed

File: tensorboard/webapp/widgets/data_table/filter_dialog.scss [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+@use '@angular/material' as mat;
+@import 'tensorboard/webapp/theme/tb_theme';
+
+.filter-dialog {
+  padding: 8px;
+  border-radius: 4px;
```
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialog.scss:20]: I've added myself a task to tackle this in the future but I would like to tackle the cleanup seperately.
https://docs.google.com/spreadsheets/d/1RBodm1hj3-njiowUeRXQe3h1GlVb70Ce1JvZw_Cp_Yc/edit?resourcekey=0-ehUesPxhJhOONsW6w_SXsA#gid=0

File: tensorboard/webapp/widgets/data_table/filter_dialog.scss (hunk 9 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialog.scss]: I don't think I checked after changing to use the new components. Apparently the padding is now `12px`.

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        includeUndefined: false,
+        possibleValues: ['foo', 'bar', 'baz', 'qaz'],
+        filterValues: ['foo', 'bar', 'baz', 'qaz'],
+      },
+    });
+    expect(
+      (await rootLoader.getAllHarnesses(MatCheckboxHarness)).length
+    ).toEqual(5); // 4 options + the include undefined checkbox
+
+    fixture.componentInstance.discreteValueFilter = 'ba';
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:130]: It seems like you are missing coverage of discreteValueKeyUp(). This seems like a good place to add that - instead of setting the discreteValueFilter variable directly can you instead type 'ba' into the box or somehow trigger the `change` event? If not, can you at least call discreteValueKeyUp() directly?

(Does using test harnesses make this easier to do?)

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    fixture.detectChanges();
+    expect(
+      (await rootLoader.getAllHarnesses(MatCheckboxHarness)).length
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:133]: I would be more confident in the test if you actually checked that it's the 'bar' and 'baz' boxes shown.

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+    expect(includeUndefinedSpy).toHaveBeenCalled();
+  });
+
+  it('dispatches event when an discrete filter value is changed', async () => {
+    const fixture = createComponent({
+      filter: {
+        type: DomainType.DISCRETE,
+        includeUndefined: false,
+        possibleValues: [2, 4, 6, 8],
+        filterValues: [2, 4, 6],
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:89]: Can we add coverage of what is rendered? Given possibleValues and filterValues, which checkboxes are rendered? (Ideally in its own test case).

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    fixture.detectChanges();
+    return fixture;
+  }
+
+  it('dispatches event when an interval filter value is changed', async () => {
+    const fixture = createComponent({
+      filter: {
+        type: DomainType.INTERVAL,
+        includeUndefined: false,
+        minValue: 6,
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:50]: Can we add coverage of what is rendered? Given minValue, maxValue, filterLowerValue, filterUpperValue, what is actually passed to (or rendered by) the tb-range-component? (Ideally in its own test case).

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+    const checkbox = await rootLoader.getHarness(
+      MatCheckboxHarness.with({label: '2'})
+    );
+    await checkbox.uncheck();
+    fixture.debugElement
+      .query(By.css('mat-checkbox label'))
+      .nativeElement.click();
+    expect(discreteFilterChangedSpy).toHaveBeenCalled();
+
+    const includeUndefinedSpy = spyOn(
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:105]: Is this its own test case 'dispatches  event when include undefined is clicked'? It doesn't seem like its part of 'dispatches event when an discrete filter value is changed'.

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    );
+
+    const rangeInput = fixture.debugElement.query(
+      By.directive(RangeInputComponent)
+    );
+    rangeInput.componentInstance.rangeValuesChanged.emit({
+      lowerValue: 7,
+      upperValue: 17,
+    });
+    expect(intervalFilterChangedSpy).toHaveBeenCalled();
```
Comment (bmd3k) [inline tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:69]: Can we check which values the spy has been called with?

**Review by bmd3k — COMMENTED** [review]
Thanks its looking pretty good. All that remains are some requests for increased test coverage and clarity.

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+      'emit'
+    );
+    const checkbox = await rootLoader.getHarness(
+      MatCheckboxHarness.with({label: '2'})
+    );
+    await checkbox.uncheck();
+    fixture.debugElement
+      .query(By.css('mat-checkbox label'))
+      .nativeElement.click();
+    expect(discreteFilterChangedSpy).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:103]: I asserted the value and few more calls.

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts (hunk 19 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:50]: I've gone ahead and checked that the range input element is shown and that the initial values are correct. I don't want this test to overlap too greatly with the discrete range input test though. `tensorboard/webapp/widgets/range_input/range_input_test.ts`

File: tensorboard/defs/defs.bzl [hunk 23]
```diff
@@ -72,8 +71,8 @@ def tf_js_binary(
         # the global level and tends to overwrite `window` functions. "iife" is
         # just a thin wrapper around "esm" (it adds 11 bytes) and doesn't
         # suffer from the same overwriting problem.
-        format="iife",
-        minify= False if dev_mode_only else True,
+        format = "iife",
```
Comment (bmd3k) [inline tensorboard/defs/defs.bzl:74]: You have unrelated changes here.

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts (hunk 17 above)
Comment (bmd3k) [reply tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:229]: This would still make me more confident. (You resolved this but I didn't see corresponding changes).

**Review by bmd3k — APPROVED** [review]
A few more comments to be addressed, please.

File: tensorboard/defs/defs.bzl (hunk 23 above)
Comment (rileyajones) [reply tensorboard/defs/defs.bzl:74]: The observer effect (aka ide autoformat) is alive and strong. Sorry I missed this.

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts (hunk 17 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:229]: I'm not sure how this has happened but this comment is on an outdated version of the file.

Clicking your comment allows me to see the code you are commenting on:
image

However, clicking the "Files Changed" tab I see a much different file (where I did address your comment).
image

image

File: tensorboard/webapp/widgets/data_table/filter_dialog_test.ts [hunk 24]
```diff
@@ -89,19 +154,36 @@ describe('filter dialog', () => {
...
+    // Unchecking an unchecked box should not trigger an event.
+    await checkbox.uncheck();
+
+    await checkbox.check();
+    checkbox = await rootLoader.getHarness(
+      MatCheckboxHarness.with({label: '4'})
+    );
+    await checkbox.uncheck();
+
+    expect(filterValues).toEqual([2, 2, 4]);
```
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/filter_dialog_test.ts:175]: This is my second draft of this. In the original I did check after each event but I thought the amount of setup required for each action made it hard to read. I'm going to put this in a loop instead but I plan to maintain the practice of checking the values in an array as I do find it more readable.
